kernel/printk/printk.c

```diff
@@ -3319,4 +3357,211 @@ void show_regs_print_info(const char *log_lvl)
 	       log_lvl, current, task_stack_page(current));
 }
 
+#ifdef CONFIG_LOGBUFFER
+
+/* Bootloader log control block address passed in on the kernel cmdline */
+phys_addr_t log_cb_phy_addr __initdata = 0;
+
+/* save requested log_cb_phy_addr since it's too early to process it */
+static int __init log_cb_phy_addr_setup(char *str)
+{
+	log_cb_phy_addr = memparse(str, &str);
+
+	return 0;
+}
+early_param("log_cb_phy_addr", log_cb_phy_addr_setup);
+
+static struct printk_log *log_from_idx_and_base(const char * const base, const u32 idx)
+{
+	struct printk_log *msg = (struct printk_log *)(base + idx);
+
+	if (!msg->len || msg->magic != LOGBUFF_MAGIC )
+		return (struct printk_log *)base;
+	return msg;
+}
+
+static u32 log_next_idx_from_idx(const char * const base, const u32 idx)
+{
+	struct printk_log const * const msg = (struct printk_log *)(base + idx);
+
+	if (!msg->len || msg->magic != LOGBUFF_MAGIC )
+		return 0;
+	return idx + msg->len;
+}
+
+static bool is_valid_external_cb ( const log_cb_t const * new_cb )
+{
+	// Validate the new control buffer
+	return (new_cb &&
+		new_cb->log_version == LOGBUFF_LOG_VERSION &&
+		new_cb->magic == LOGBUFF_MAGIC &&
+		new_cb->log_cb_pad_length ==
+		LOGBUFF_CB_PAD_LENGTH &&
+		new_cb->log_cb_t_size == sizeof(log_cb_t) &&
+		new_cb->log_header_size ==
+		sizeof(struct printk_log) &&
+		new_cb->log_cb_t_size <= new_cb->log_cb_pad_length &&
+		new_cb->log_first_idx <= new_cb->log_length &&
+		new_cb->log_next_idx <= new_cb->log_length &&
+		new_cb->syslog_idx <= new_cb->log_length &&
+		new_cb->console_idx <= new_cb->log_length &&
+		new_cb->clear_idx <= new_cb->log_length &&
+		new_cb->log_phy_addr != 0);
+}
+
+// TODO: Release memory replaced by external buffer
+void __init setup_ext_logbuff(void)
+{
+	struct printk_log *nextblmsg, *currkmsg;
+	u64 currkmsg_seq;
+	u32 currkmsg_idx, offset_idx;
+	log_cb_t *new_cb;
+	unsigned long flags;
+
+	new_cb = (log_cb_t*) log_cb_phy_addr;
+
+	/* Check new CB pointer. Ignore the external buffer if NULL. */
+	if ( ! new_cb )
+	{
```

Comment:
This is the first place I noticed it, the convention I've mostly seen throughout the kernel is to cuddle braces with the conditional.  I don't like it, but it's not my kernel.  :-)